helm: kvstoremesh with external etcd

calcium6 opened the pull request:
When running clustermesh-apiserver, there should be an option to enable and disable containers that are not needed.
For example the etcd and apiserver containers are not necessary when running kvstoremesh with an external kvstore.

This commit adds the options to disable these containers.

giorio94 commented on examples/kvstoremesh/kvstoremesh.yaml:
I'd suggest to add a note to the documentation on how to deploy kvstoremesh in this context, rather than adding an example here. Examples tend to be easily missed and get stale over time, as well as require a-priory knowledge about the specific goal, as lacking the accompanying description.

calcium6 committed:
helm: kvstoremesh with external etcd

have the option to use external etcd with kvstoremesh

Signed-off-by: Jiri Vanura <jiri.vanura@firma.seznam.cz>

giorio94 commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -2938,6 +2938,8 @@ clustermesh:
   # Supported values are 255 and 511.
   maxConnectedClusters: 255
+  # -- Use external KVStore for KVStoreMesh.
+  kvstoremeshExternalKVStore: false
```
The dedicated `clustermesh.kvstoremesh` section seems a better fit to me. Additionally, I'd clarify in the comment that the target etcd instance is the one configured via `etcd.endpoints`, as not obvious otherwise.

Do you have any opinions on:

> If we decide that kvstoremesh+sidecar etcd is a valid use case, then we may have a mode flag that allows selecting among the three possible alternatives.

Mostly to decide whether to generalize the setting to potentially enable this combination in the future (it doesn't need to be as part of this PR, considering that it requires a bit of extra work). Still, I'd rather prefer to pick a future-proof setting from the beginning if possible, to avoid having to deal with backward-compatible changes.

calcium6 replied:
Yes, I also think it would be better to add a mode flag to select from the three options, I'll do that.

balous replied:
I believe this comment has already been resolved by the original submitter:

https://github.com/cilium/cilium/pull/34174/files#diff-0ea42ad21164b19bec1732225e254d3096d1e4040481c00053669287d81015feR3055

giorio94 commented on install/kubernetes/cilium/templates/validate.yaml:
```diff
@@ -153,7 +153,7 @@
 {{/* validate clustermesh-apiserver */}}
 {{- if .Values.clustermesh.useAPIServer }}
-  {{- if and (ne .Values.identityAllocationMode "crd") (ne .Values.identityAllocationMode "doublewrite-readkvstore") (ne .Values.identityAllocationMode "doublewrite-readcrd") }}
+  {{- if and (ne .Values.identityAllocationMode "crd") (ne .Values.identityAllocationMode "doublewrite-readkvstore") (ne .Values.identityAllocationMode "doublewrite-readcrd") (not .Values.clustermesh.kvstoremeshExternalKVStore) }}
```
I'd suggest adding a similar validation to ensure that the external kvstore mode can only be enabled if `etcd.enabled` is set to true, and `etcd.endpoints` is not empty, to avoid surprises (given the current implementation). Similarly, I'd suggest validating that `kvstoremesh.enabled=true` when `useAPIServer=true` and `kvstoremeshExternalKVStore=true`, otherwise the resulting deployment is invalid. Finally, it should be forbidden to enable `externalWorkloads.enabled=true` if `kvstoremeshExternalKVStore=true`.

balous replied:
I believe this comment has already been resolved: https://github.com/cilium/cilium/pull/34174/files#diff-63d6e1fdb929a5058c43790c57f422863b269e45c2be10eb6564ca51fe1a0f4eR171

giorio94 commented on Documentation/network/clustermesh/clustermesh.rst:
```diff
@@ -147,6 +147,11 @@ Example install using the Cilium CLI:
+.. note::
+
+  If you are planning on using external etcd with your cluster, you have to set the value of the Helm option ``clustermesh.kvstoremeshExternalKVStore`` 
+  to ``true`` for the KVStoreMesh to run with the external etcd instead of creating a new etcd instance upon creation.
```
Nit. We typically refer to KVStoreMesh, rather than *the* KVStoreMesh.

giorio94 commented on install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml:
The creation of the clustermesh-apiserver specific `ClusterRole` and `ClusterRoleBinding` should be disabled when running only KVStoreMesh, as not required. Similarly, the creation of the `clustermesh-apiserver` Service should be disabled, and the `clustermesh-apiserver-metrics` Service/ServiceMonitor should be adapted to disable all but KVStoreMesh-related entries. Finally, the creation of the `clustermesh-remote-users` ConfigMap and of the different TLS-related secrets should be most likely disabled as well.

balous replied:
I believe this coment has already been resolved by the original submiter.

giorio94 commented on install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml:
```diff
@@ -98,7 +99,9 @@ spec:
         {{- end }}
+      {{- end }}
       containers:
+      {{- if not .Values.clustermesh.kvstoremeshExternalKVStore}}
```
There are a bunch of `volumes` configured for this deployment that should be skipped when running in KVStoreMesh-only mode, as not mounted in that case.

balous replied:
I believe this comment has already beed resolved: 
- https://github.com/cilium/cilium/pull/34174/files#diff-c85fb06a8dab414080c30e0cf3361b9baade1b357dd2d90e1a1797be2dbdcf31R358
- https://github.com/cilium/cilium/pull/34174/files#diff-c85fb06a8dab414080c30e0cf3361b9baade1b357dd2d90e1a1797be2dbdcf31R336

giorio94 commented on install/kubernetes/cilium/templates/clustermesh-config/clustermesh-secret.yaml:
```diff
@@ -11,7 +11,7 @@ metadata:
 data:
   {{- $kvstoremesh := and .Values.clustermesh.useAPIServer .Values.clustermesh.apiserver.kvstoremesh.enabled }}
-  {{- $override := ternary (printf "https://clustermesh-apiserver.%s.svc:2379" .Release.Namespace) "" $kvstoremesh }}
+  {{- $override := ternary (join "\n- " $.Values.etcd.endpoints) (ternary (printf "https://clustermesh-apiserver.%s.svc:2379" .Release.Namespace) "" $kvstoremesh) .Values.clustermesh.kvstoremeshExternalKVStore }}
```
When external KVStore mode is enabled, the Cilium agents appear to default using the etcd endpoints configured via `.Values.etcd.endpoints`. However, this change does not seem to account for the handling of the TLS certificates, accounting for `.Values.etcd.ssl`. Similarly, neither the endpoints nor the certificates seem to be automatically configured in the `kvstoremesh` container, which would not work out of the box.

giorio94 commented on install/kubernetes/cilium/templates/validate.yaml:
```diff
@@ -153,7 +153,7 @@
+  {{- if and (ne .Values.identityAllocationMode "crd") (ne .Values.identityAllocationMode "doublewrite-readkvstore") (ne .Values.identityAllocationMode "doublewrite-readcrd") (not .Values.clustermesh.kvstoremeshExternalKVStore) }}
     {{ fail (printf "The clustermesh-apiserver cannot be enabled in combination with .Values.identityAllocationMode=%s. To establish a Cluster Mesh, directly configure the parameters to access the remote kvstore through .Values.clustermesh.config" .Values.identityAllocationMode ) }}
   {{- end }}
   {{- if .Values.disableEndpointCRD }}
```
The `disableEndpointCRD` validation can be skipped if `kvstoremeshExternalKVStore=true`.

balous replied:
I believe this comment has already been resolved: https://github.com/cilium/cilium/pull/34174/files#diff-63d6e1fdb929a5058c43790c57f422863b269e45c2be10eb6564ca51fe1a0f4eR155

giorio94 reviewed (commented):
Thanks for the changes.

I've left a few additional comments inline. Overall, the main concern I have about the current proposal is that it would work in a very specific use-case only (and with specific settings, which are also not documented), while leading to a most likely broken configuration otherwise. I personally think that it should be generalized a bit so that it works out of the box in most common scenarios, as well as the validation should be extended to provide safe-guards and prevent users from enabling unsupported configurations. My comments mostly go in that direction. WDYT?

giorio94 reviewed (changes requested):
(Requesting changes as per above comments)

calcium6 committed:
helm: validation
changed the flag name clustermesh.kvstoremeshExternalKVStore to clustermesh.apiserver.kvstoremesh.useExternalKVStore

Signed-off-by: Jiri Vanura <jiri.vanura@firma.seznam.cz>

calcium6 committed:
helm: kvstore modes for kvstoremesh
changed the flag clustermesh.apiserver.kvstoremesh.useExternalKVStore to clustermesh.apiserver.kvstoremesh.kvstoreMode

Signed-off-by: Jiri Vanura <jiri.vanura@firma.seznam.cz>

balous commented on Documentation/network/clustermesh/clustermesh.rst:
```diff
@@ -147,6 +147,11 @@ Example install using the Cilium CLI:
   cilium install --set cluster.name=$CLUSTER1 --set cluster.id=1 --context $CLUSTER1
   cilium install --set cluster.name=$CLUSTER2 --set cluster.id=2 --context $CLUSTER2
 
+.. note::
```
I believe this comment has already been resolved by the original submitter:

https://github.com/cilium/cilium/pull/34174/files#diff-8c6c2ac85134f66ed70f9aca62524e18ac4b8003f7d41101521d023e491f3f82R199
